eng/pipelines/templates/steps/build-test.yml

```diff
@@ -35,6 +35,7 @@ steps:
       python -m pip install pip==20.3.3
       python -m pip install wheel==0.37.0
       python -m pip install setuptools==58.3.0
+      python -m pip install docker==4.3.1
```

Comment:
What are you trying to accomplish here? This is ensuring that `docker` gets installed _before_ invoking the test environments in each, but it doesn't actually affect what's being installed in the running tox environments.

Reply:
Enabling an E2E test. When running the test on local setup it works as I have docker running. However, when this is run as part of PR gate getting this error.
DockerEngineNotAvailableError: Please make sure Docker Engine is installed and running. 
Any idea how to start the docker engine on the build agent?

Reply:
> Enabling an E2E test. When running the test on local setup it works as I have docker running. However, when this is run as part of PR gate getting this error. DockerEngineNotAvailableError: Please make sure Docker Engine is installed and running. Any idea how to start the docker engine on the build agent?

Docker engine's run inconsistently on the build agents.

1. Windows is running the `windows` docker only, no way to change that I'm aware of.
2. Linux is running `linux` docker only. No way to change that I'm aware of.
3. Mac isn't licensed to run docker in devops anymore.

I opened your error

```
            raise Exception(
                "Please install docker in the current python environment with `pip install docker` and try again."
>           ) from e
E           Exception: Please install docker in the current python environment with `pip install docker` and try again.

```

I don't know why this works for the `ubuntu` agent, but I'm betting it's because you're escaping out of the current virtual environment and using the global `docker` from whatever the invoking python version is. 

Your CI runs are actually invoked within parallel virtual environments by `tox`. The PR you have now is adding `docker` to the _invoking_ python environment, but when `tox` spins up it shouldn't actually have access to that version. If it _does_, it's because it's on the PATH globally and you're not actually using the python executable from within the tox test env. That's probably not a good thing.

Reply:
I recommend you add `docker` to your `dev_requirements` instead of where you're adding it. That will have the affect you're looking for.

However, you'll probably want to ensure there is a platform qualifier on the req so it only installs for `windows` and `linux`.

Reply:
Thanks for sharing pointers.